oauth2/fosite_store_sql.go

```diff
@@ -355,3 +363,53 @@ func (s *FositeSQLStore) FlushInactiveAccessTokens(ctx context.Context, notAfter
 
 	return nil
 }
+
+// internal convenience interface to keep storage methods DRY otherwise there will be hella lot of copy-pasta
+// note sure if this is the best way to go about this tho ¯\_(ツ)_/¯
+type operatorExtContext interface {
+	sqlx.ExtContext
+	GetContext(ctx context.Context, dest interface{}, query string, args ...interface{}) error
+	NamedExecContext(ctx context.Context, query string, arg interface{}) (sql.Result, error)
+}
+
+// helper used to determine whether we are dealing with a transaction or not. name is kinda shitty?
+func (s *FositeSQLStore) operator(ctx context.Context) (operatorExtContext) {
+	if txn, ok := ctx.Value(transactionKey).(*sqlx.Tx); ok {
+		s.L.Infof("AMIR -------------------------- OPERATOR IS TXN - %+v", txn)
+		return txn
+	} else {
+		return s.DB
+	}
+}
+
+func (s *FositeSQLStore) BeginTX(ctx context.Context) (context.Context, error) {
+	if txn, err := s.DB.BeginTxx(ctx, nil); err != nil {
+		s.L.Errorf("AMIR -------------------------- BeginTX TRANSACTION ERROR: %+v", err)
+		return ctx, err
+	} else {
+		s.L.Infof("AMIR -------------------------- BeginTX SUCCESS: %+v", txn)
+		return context.WithValue(ctx, transactionKey, txn), nil
+	}
+}
+
+func (s *FositeSQLStore) Commit(ctx context.Context) error {
+	if txn, ok := ctx.Value(transactionKey).(*sqlx.Tx); ok {
+		s.L.Infof("AMIR -------------------------- COMMITING TRANSACTION")
+		return txn.Commit()
+	} else {
+		// TODO: probably don't want to panic here. Although it if this method is called by fosite
+		// it's because you said you support transactions! So...are you lying?
+		panic("no transaction available in context - wtf not cool man.")
```

Comment:
@aeneasr 👀

Reply:
I think we can handle this gracefully with a proper error, in case there is a bug it won't bust the whole system but only certain transactions!